### PR TITLE
Improving output cache performance

### PR DESCRIPTION
**Pull request by sebastienros**
- Preventing multiple requests from requesting the same cache item concurrently
- Using TaskCompletionSource to synchronize workers instead of locks

### CODE REVIEW

**Comment by sebastienros** [comment]
Tests:
- Added some delay in the view itself to ensure a single thread at a time computes the value.
- Added some delay in the `DefaultOutputCacheStorage` to ensure a single thread at a time requests a cached item.
- Verified grace time behavior.

I used DEBUG log to ensure all worked correctly, which is why I added more log everywhere.

**Comment by sebastienros** [comment]
Just in case you didn't see this one /cc  @DanielStolt @Piedone

File: src/Orchard.Web/Modules/Orchard.OutputCache/Filters/OutputCacheFilter.cs [hunk 1]
```diff
@@ -117,53 +121,44 @@ public class OutputCacheFilter : FilterProvider, IActionFilter, IResultFilter, I
                 if (allowServeFromCache && cacheItem != null) {
 
                     Logger.Debug("Item '{0}' was found in cache.", _cacheKey);
+                    
+                    if(cacheItem.IsValid(_now)) {
```
Comment (Piedone) [inline src/Orchard.Web/Modules/Orchard.OutputCache/Filters/OutputCacheFilter.cs]: Space is needed after "if".

File: src/Orchard.Web/Modules/Orchard.OutputCache/Filters/OutputCacheFilter.cs [hunk 2]
```diff
@@ -117,53 +121,44 @@ public class OutputCacheFilter : FilterProvider, IActionFilter, IResultFilter, I
...
+                    else {
+                        Logger.Debug("Cached item is invalid: {0}", _cacheKey);
 
-                    // Is the cached item in its grace period?
-                    if (cacheItem.IsInGracePeriod(_now)) {
-
-                        // Render the content unless another request is already doing so.
-                        if (Monitor.TryEnter(_cacheKey)) {
+                        // If the cached item is in its grace period 
+                        // and there are no current renderer, this request renders the new content
```
Comment (Piedone) [inline src/Orchard.Web/Modules/Orchard.OutputCache/Filters/OutputCacheFilter.cs]: Nitpicking again but the comment needs punctuation at the end and there is a grammatical error: either "are ... renderers" or "is ... renderer". Judging from the context the latter one.

File: src/Orchard.Web/Modules/Orchard.OutputCache/Filters/OutputCacheFilter.cs [hunk 3]
```diff
@@ -602,15 +613,51 @@ public class OutputCacheFilter : FilterProvider, IActionFilter, IResultFilter, I
...
+                Logger.Debug("Fetching cache item: {0}", _cacheKey);
+
+                try {
+                    var cacheItem = _cacheStorageProvider.GetCacheItem(key);
+                    _state.Cachers.TryRemove(_cacheKey, out tcs);
+                    tcs.SetResult(cacheItem);
+                    return cacheItem;
+                }
+                catch (Exception e) {
+                    Logger.Error(e, "An unexpected error occured while reading a cache entry");
```
Comment (Piedone) [inline src/Orchard.Web/Modules/Orchard.OutputCache/Filters/OutputCacheFilter.cs]: If you catch the base Exception at least add an `e.IsFatal()` check. BTW aren't we rather conventionally using the variable name `ex` for exceptions, `e` being something common for event object in event handlers?

**Comment by Piedone** [comment]
This looks like a lot of complexity added. Do you think it's really worth preventing cases of duplicate work when the same uncached page is requested by multiple clients concurrently? Especially that this is a local lease, so it won't prevent the same across multiple server nodes. Does it indeed increase throughput of an actual application running somewhere, i.e. is there evidence of these changes being worth it?

**Comment by sebastienros** [comment]
There is evidence, we just deployed it to scottgu's blog in prod because with Redis it just doesn't work without it.
The code is more complex, but also removes the locking that was used before.

File: src/Orchard.Web/Modules/Orchard.OutputCache/Filters/OutputCacheFilter.cs (hunk 1 above)
Comment (sebastienros) [inline src/Orchard.Web/Modules/Orchard.OutputCache/Filters/OutputCacheFilter.cs]: I don't know what I changed, all my `if` are not formatted automatically anymore :/

**Comment by sebastienros** [comment]
I will explain the scenario behind this change tomorrow, we'll see how much sense it makes.

**Comment by DanielStolt** [comment]
@Piedone You write "uncached page" - this makes me suspect either you or I have misunderstood the purpose of the PR. As I understood, this is about preventing unnecessary load on the cache storage service, i.e. cases when the same *cached* page is requested by multiple clients concurrently.

**Comment by DanielStolt** [comment]
I've looked this over quickly, and I while I believe the initial idea is indeed a valuable functional improvement, I have reservations about the implementation. I will review it some more in the coming days but here are some initial thoughts.

- Stop removing the trailing periods in comments...! ;)

- I think this additional synchronization should be implemented by the storage provider, not the output cache filter, because a) it keeps the already complex filter a little simpler and cleaner, and b) the need in the first place is specific to a particular storage provider where the load on the storage is expensive - makes little sense for the default provider which is only in-memory. I do understand that you're trying to kill two birds with one stone and using the `TaskCompletionSource` mechanism to also implement the synchronization for *rendering* an item, but I think you are probably underestimating how tricky that logic is to get 100% correct, and your change is so big and sweeping that it's very likely you've broken some scenarios. I can't make sense of your new logic, but perhaps it would help if you walk us through it.

- I think you are doing too much in the `BeginRenderItem` method, this method was only meant to initialize/set up the rendering.

- The `IOutputCacheFilterState` seems like a completely unnecessary and very arbitrary abstraction to me. What is the purpose? How could anyone possibly implement it any different? Instead it should be part of the responsibility and implementation of the `OutputCacheFilter` implementation IMO.

**Comment by sebastienros** [comment]
I must have installed a plugin that messes with the comments ;) Fortunately we are using PRs.

> I think this additional synchronization should be implemented by the storage provider

I see why you would think that, my concern is that the filter itself is broken into 4 specific events, and I didn't see any obvious solution to do this. Maybe splitting the rendering and fetching would do it. I will try to separate the concerns and see if there is any dependency between the two. The rendering lock mecanism would remain in the filter, and the cache synchronization would be moved in the storage implementation on a case by case basis.

> you are doing too much in the BeginRenderItem

I should probably have renamed it, it's definitely not the same purpose with that change.

> The IOutputCacheFilterState seems like a completely unnecessary

I needed some state that is shared across all requests of a tenant. The only way I could think of was a Singleton dependency. I didn't want to use the `IMemoryCache` for that, abusing the interface.

**Comment by DanielStolt** [comment]
OK, I understand - the singleton lifetime is a valid reason for the abstraction.

**Comment by Piedone** [comment]
@DanielStolt yes I misunderstood, thanks for the clarification.

**Pull request closed without merging**